protocols/dcutr/src/behaviour_impl.rs

```diff
@@ -282,33 +279,32 @@ impl NetworkBehaviour for Behaviour {
         addr: &Multiaddr,
         role_override: Endpoint,
     ) -> Result<THandler<Self>, ConnectionDenied> {
-        match self
-            .outgoing_direct_connection_attempts
-            .remove(&(connection_id, peer))
+        if is_relayed(addr) {
+            return Ok(Either::Left(handler::relayed::Handler::new(
+                ConnectedPoint::Dialer {
+                    address: addr.clone(),
+                    role_override,
+                },
+            ))); // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
+        }
+
+        if let Some(&relayed_connection_id) = self.direct_to_relayed_connections.get(&connection_id)
         {
-            None => {
-                let handler = if is_relayed(addr) {
-                    Either::Left(handler::relayed::Handler::new(ConnectedPoint::Dialer {
-                        address: addr.clone(),
-                        role_override,
-                    })) // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
-                } else {
-                    Either::Right(Either::Right(dummy::ConnectionHandler))
-                };
-
-                Ok(handler)
-            }
-            Some(_) => {
+            if role_override == Endpoint::Listener {
                 assert!(
```

Comment:
I think this `assert` operates under the assumption that no other `NetworkBehaviour` implementation emits a `ToSwarm::Dial` with `DialOpts::role_override`. I think it is fine to make that assumption. Just want to call it out here.

Reply:
Never mind, given that it is nested within the `if ... = self.direct_to_relayed_connections` we can be sure that this connection is one of ours.